src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -351,109 +360,151 @@ internal override async ValueTask<int> ReadAsync(Memory<byte> destination, Cance
                 throw new InvalidOperationException(SR.net_quic_reading_notallowed);
             }
 
-            if (cancellationToken.IsCancellationRequested)
-            {
-                lock (_state)
-                {
-                    if (_state.ReadState == ReadState.None)
-                    {
-                        _state.ReadState = ReadState.Aborted;
-                    }
-                }
-
-                throw new OperationCanceledException(cancellationToken);
-            }
-
             if (NetEventSource.Log.IsEnabled())
             {
                 NetEventSource.Info(_state, $"[Stream#{_state.GetHashCode()}] reading into Memory of '{destination.Length}' bytes.");
             }
 
+            ReadState readState;
+            long abortError = -1;
+            bool canceledSynchronously = false;
+
             lock (_state)
             {
-                if (_state.ReadState == ReadState.ReadsCompleted)
-                {
-                    return 0;
-                }
-                else if (_state.ReadState == ReadState.Aborted)
+                readState = _state.ReadState;
+                abortError = _state.ReadErrorCode;
+
+                if (readState != ReadState.PendingRead && cancellationToken.IsCancellationRequested)
```

Comment:
Why not in `PendingRead`?
Do we somehow guard against parallel reads? Does this PR contain something that prevents that?
EDIT: I see it in the switch after.